### ISSUE TITLE
Modding with GFDM old series will not work properly

### ISSUE THREAD

**Issue by JamesLewisLiu** [mon/ifs_layeredfs#5]
Though unpack is working now, modding textures is not working properly while I'm downloading remaining games' data , with modded texture game will blank out on modded textures like this
<img width="480" alt="image" src="https://github.com/mon/ifs_layeredfs/assets/17558260/9abefe9a-e4a4-4787-9f47-a44f7787615a">
This is unpacked and converted with gitadora-texbintool.exe data
![image](https://github.com/mon/ifs_layeredfs/assets/17558260/e6f0236e-2efc-4917-930a-b189369ba1d7)
And I mod like this
![image](https://github.com/mon/ifs_layeredfs/assets/17558260/f6e1eb83-5ad1-4982-8516-82f51c6f52d3)
Here's two logs
[ifs_hook_unpkfs_and_mod.log](https://github.com/mon/ifs_layeredfs/files/15151551/ifs_hook_unpkfs_and_mod.log)
[ifs_hook.log](https://github.com/mon/ifs_layeredfs/files/15151552/ifs_hook.log)

**Comment by mon** [comment]
Can you send both the original `tex_gf_game2.bin` as well as `./data_mods/_cache/product/d3/model/tex_gf_game2.bin`

**Comment by JamesLewisLiu** [comment]
_cached
[_cached.zip](https://github.com/mon/ifs_layeredfs/files/15156118/_cached.zip)
unpacked
[unpacked one.zip](https://github.com/mon/ifs_layeredfs/files/15156131/unpacked.one.zip)

**Comment by mon** [comment]
I'm starting to build an idea as to what causes this. If you repack using gitadora-texbintool (instead of letting layeredfs do it) does it still break?

**Comment by JamesLewisLiu** [comment]
Tried using bin as mod directly and still no luck
![image](https://github.com/mon/ifs_layeredfs/assets/17558260/892aaf50-08f2-4f72-a9d5-59c3f97c9733)
Here's log
[ifs_hook.log](https://github.com/mon/ifs_layeredfs/files/15172988/ifs_hook.log)

**Comment by mon** [comment]
And what is the expected result here? The image spans the entire screen?

**Comment by JamesLewisLiu** [comment]
Sorry for late reply
Different stages have different backgrounds(and two PREMIUM ENCORE have two special one), I just want to test change every background to the A.DOGMA Special one as the following video shows
https://github.com/mon/ifs_layeredfs/assets/17558260/b5ef13d9-4571-428d-ad6e-5cda7dd98b31
And with the testmod applying now not only the background but also almost every texture blank out as this screenshot shows ![image](https://private-user-images.githubusercontent.com/17558260/327036079-892aaf50-08f2-4f72-a9d5-59c3f97c9733.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUwMTE3MjgsIm5iZiI6MTcxNTAxMTQyOCwicGF0aCI6Ii8xNzU1ODI2MC8zMjcwMzYwNzktODkyYWFmNTAtMDhmMi00ZjcyLWE5ZDUtNTljM2Y5N2M5NzMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA2VDE2MDM0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2ZGU5NTdlZWQwNThhY2ViNWFlMmJjYmQxNzNlZWQ3YTk3Zjc0YmViNGQxNTBhMGIyMjYyODMxZmNmZDNlYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V0ilWU1_3oh2OnXAwVBG2vXIfHZy-fzKKQFkhZoGtdM)

**Comment by JamesLewisLiu** [comment]
Today I was wondering is modding XG possible
and I test with a BGA first, it works fine like this
![image](https://github.com/mon/ifs_layeredfs/assets/17558260/3f9b0c82-48be-4610-aae4-fad0a12cb95d)
Then I tried pick a random texture and mod it however the game seems don't like it and return a HDD Game Error like this
![image](https://github.com/mon/ifs_layeredfs/assets/17558260/0fba4bb6-e163-44fb-98ab-dfaf29995b4b)
Here's log
[ifs_hook.log](https://github.com/mon/ifs_layeredfs/files/15275674/ifs_hook.log)
Here's texture file
[texture.zip](https://github.com/mon/ifs_layeredfs/files/15275817/texture.zip)

**Comment by mon** [comment]
I'm continuing to look into this, and I think I might have a reason for the crashes. Just to confirm some things:

> Tried using bin as mod directly and still no luck

This was repacked using gitadora-texbintool? If so, then I have a bit of work to do to fix this😅

**Comment by mon** [comment]
If possible, for your HDD ERROR case, please send the original + modded .bin files for me to inspect. If they show the same pattern as the other non-working file, then I know what to try.

**Comment by JamesLewisLiu** [comment]
> This was repacked using gitadora-texbintool? If so, then I have a bit of work to do to fix this😅

Yes, I don't know if there are any other tools to do that thing.


> If possible, for your HDD ERROR case, please send the original + modded .bin files for me to inspect. If they show the same pattern as the other non-working file, then I know what to try.

Since it's been a while I test both V5 and XG again and here is my mod and log files, hope these can help you
btw I watched it for a while and the XG1 instantly reported that error when ifs_layeredfs tried loading mod files
[V5.zip](https://github.com/mon/ifs_layeredfs/files/15286453/V5.zip)
[XG1.zip](https://github.com/mon/ifs_layeredfs/files/15286454/XG1.zip)

**Comment by mon** [comment]
Oh dear, my pattern was wrong... I was thinking "maybe it's DXT5 compression instead of BGRA", but the XG1 file looks practically identical:
![image](https://github.com/mon/ifs_layeredfs/assets/3398021/94ced5e4-d7db-4c57-b068-8689897533ca)

Though, since V5 doesn't HDD ERROR but instead just has a black image, DXT5 might help there as well. But it's a bit more effort, so I want to solve XG1 first.

Please give XG1 another launch with this version (always_devmode_verbose_and_logs_to_file_and_pkfs_unpack) - it won't fix anything, but it'll make some more logs around failed calls to pkfs_open, which I think might be the reason for the HDD ERROR (it's something I fixed in jubeat recently).
[ifs_layeredfs_3.4_RC3.zip](https://github.com/mon/ifs_layeredfs/files/15294528/ifs_layeredfs_3.4_RC3.zip)

Also thank you for your continued assistance, I really appreciate it!

**Comment by JamesLewisLiu** [comment]
It still reports HDD READ ERROR, here's log
[ifs_hook.log](https://github.com/mon/ifs_layeredfs/files/15296761/ifs_hook.log)

**Comment by mon** [comment]
Well, this is exactly the log I needed!

```
M:pkfs_fs_open(./data_mods/_cache/product/d3/model/tex_gf_common.bin) failed in call_real
```

...unfortunately, debugging this is going to be tricky. This type of call has never failed before. I can't actually find a copy of Gitadora XG. I will ask around.

**Comment by mon** [comment]
Oh, I think I finally remembered what was causing this... The same issue happens with gfdm 5...

Old GFDM uses launcherx/gfdmhook, which isn't as polished as mainline bemanitools/spice, so this hasn't been fixed yet.
It has an issue where it uses the "wrong" working directory, because the game has folders structured as so:
```
bin/
  data_mods/ <- Possible location 1
  ...
data/
  ...
data_mods/ <- Possible location 2
  ...
```
When layeredfs tries to access a file, it can only see files in `bin/`, as that's where the exe is run from, so it needs to be in "Possible location 1". But when the game's pkfs layer tries to open a file, it looks in the top level, i.e. "Possible location 2".

As far as I can tell based on newer games, the "correct" way is to have the working dir set at the toplevel, and launch modules from there, instead of retargeting the filesystem. It then also requires intercepting `avs_boot` and changing the value of `/config/fs/root/device` from `../` to `./`.
bemanitools already does this for iidxhook4, iidxhook5, and jbhook1.

## To fix it for now
- Move all of the game files out of `bin` and into the same folder as `data` and `data_mods`
- Edit `boot.dll` to change the working dir from `../` to `./`
![image](https://github.com/mon/ifs_layeredfs/assets/3398021/c41c6d00-afd0-4084-9139-bdcecada1819)


I can't think of a clean way to fix this in layeredfs. Will take any suggestions.

**Comment by JamesLewisLiu** [comment]
Well the gfdmhook have not been update for 5 years
So I use `mklink /J ..\data_mods data_mods` to create a directory junction and now it works fine
![image](https://github.com/mon/ifs_layeredfs/assets/17558260/e8a4f92a-3438-449a-9c5c-4538e5090cb4)
Since this game is not that popular may be this can be a workaround

**Comment by mon** [comment]
Do you think an addition to the README would be enough? I think if I'm clever I can probably write some sort of "is the filesystem a bit funky" test, but it might be a little work.

**Comment by JamesLewisLiu** [comment]
> Do you think an addition to the README would be enough? I think if I'm clever I can probably write some sort of "is the filesystem a bit funky" test, but it might be a little work.

I think thats' enough but with a big WARNING, filesystem link is quite tricky and may cause data loss.

**Comment by JamesLewisLiu** [comment]
![image](https://github.com/mon/ifs_layeredfs/assets/17558260/aa568f5a-4731-49a4-8777-325387e7aea4)
Just wonder maybe V4-V8 have same "wrong" directory issue and I create a link in V5 like before and it works

**Comment by mon** [comment]
I've released 3.4, and added information about the path remapping to the README. I hope that this closes this issue!
